test: Run mempool_updatefromblock even with wallet disabled

reemuru opened the pull request:
This is a PR proposed in #20078

This PR enables one more of the non-wallet functional tests (mempool_updatefromblock.py) to be run even with the Bitcoin Core wallet disabled by using the new MiniWallet instead.

Overview:
*  Refactoring and removal of unnecessary imports and arguments to transaction_graph_test()
*  Improve logging
*  Decrease test execution time by ~70% (39s => 12s)

Before:
```bash
[USER@SERVER functional]$ time ./mempool_updatefromblock.py 
2021-05-19T21:02:10.809000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_42xs1s0g
2021-05-19T21:02:11.154000Z TestFramework (INFO): Creating 100 transactions...
2021-05-19T21:02:15.741000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:02:15.756000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:02:20.097000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:02:20.115000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:02:24.687000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:02:24.705000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:02:28.774000Z TestFramework (INFO): The last batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:02:29.160000Z TestFramework (INFO): All of the recently mined transactions have been re-added into the mempool in 0.3855452537536621 seconds.
2021-05-19T21:02:29.161000Z TestFramework (INFO): Checking descendants/ancestors properties of all of the in-mempool transactions...
2021-05-19T21:02:50.455000Z TestFramework (INFO): Stopping nodes
2021-05-19T21:02:50.508000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_42xs1s0g on exit
2021-05-19T21:02:50.508000Z TestFramework (INFO): Tests successful

real    0m39.895s
user    0m34.627s
sys     0m2.474s
```

After:
```bash
[USER@SERVER functional]$ time ./mempool_updatefromblock.py 
2021-05-19T21:08:17.190000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test__3fuskkh
2021-05-19T21:08:17.493000Z TestFramework (INFO): Creating 100 transactions...
2021-05-19T21:08:17.634000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:08:17.686000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:08:17.771000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:08:17.820000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:08:17.896000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:08:17.943000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:08:18.009000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2021-05-19T21:08:18.058000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2021-05-19T21:08:18.060000Z TestFramework (INFO): Mempool size pre-invalidation: 0
2021-05-19T21:08:18.080000Z TestFramework (INFO): Mempool size post-invalidation: 100
2021-05-19T21:08:18.082000Z TestFramework (INFO): All of the recently mined transactions have been re-added into the mempool in 0.0179746150970459 seconds.
2021-05-19T21:08:18.082000Z TestFramework (INFO): Checking descendants/ancestors properties of all of the in-mempool transactions...
2021-05-19T21:08:29.441000Z TestFramework (INFO): Stopping nodes
2021-05-19T21:08:29.496000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test__3fuskkh on exit
2021-05-19T21:08:29.496000Z TestFramework (INFO): Tests successful

real    0m12.465s
user    0m5.843s
sys     0m0.800s
```

Commit:
test: Run mempool_updatefromblock even with wallet disabled

Enables the mempool_updatefromblock non-wallet functional test
to be run even with the Bitcoin Core wallet disabled by using
the new MiniWallet instead.

- Refactoring and removal of unnecessary imports and arguments to
  transaction_graph_test()
- Improve logging
- Decrease test execution time by ~70% (39s => 12s)

DariusParvin commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +35,52 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
-                if not first_block_hash:
-                    first_block_hash = block_hash
-                assert_equal(len(self.nodes[0].getrawmempool()), 0)
+        self.wallet.generate(1)
```
`scan_blocks` is slightly quicker than `generate`, so instead we can use
```suggestion
        self.wallet.scan_blocks(start=76, num=1)
```
(76 is the first block where the coinbase is mined to the miniwallet address)

reemuru replied:
Ah, I see. I was wondering how to implement the scan_blocks() but wasn't sure. Thanks! ac8b08b

DariusParvin commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +35,52 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
+            txs.append(self.wallet.send_self_transfer(from_node=node))
+            tx_size.append(96)
+            if (i+1) % n_tx_to_mine == 0:
+                self.log.info('The batch of {} transactions has been accepted into the mempool.'
```
Would it be useful to have this in the line above?
```suggestion                
                assert_equal(len(node.getrawmempool()), n_tx_to_mine)
```

reemuru replied:
Shoots, I thought I had that but it was just assert for mempool after mining a block. Updated ac8b08b

DariusParvin commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +35,52 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
         self.log.info('Checking descendants/ancestors properties of all of the in-mempool transactions...')
-        for k, tx in enumerate(tx_id):
+        k = 0
+        for tx in txs:
```
I'm curious if there's any advantage to incrementing k manually rather than using `enumerate`? Otherwise `enumerate` seems more concise.
```suggestion
        for k, tx in enumerate(txs):
```

reemuru replied:
I checked and there is no significant difference in test execution time. Using enumerate ac8b08b

DariusParvin reviewed (commented):
obvious concept ACK. 
Great work! I left some suggestions/questions

reemuru commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +35,52 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
+            tx_size.append(96)
+            if (i+1) % n_tx_to_mine == 0:
+                self.log.info('The batch of {} transactions has been accepted into the mempool.'
+                .format(len(node.getrawmempool())))
```
Nice catch! Yes will update.

reemuru commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +35,52 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
-                end = time.time()
-                assert_equal(len(self.nodes[0].getrawmempool()), size)
-                self.log.info('All of the recently mined transactions have been re-added into the mempool in {} seconds.'.format(end - start))
+                self.log.info('{}'.format(block_ids))
```
I think I was using this for debugging and set to info by mistake. Should remove this to reduce clutter on the console for sure.

reemuru commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +35,52 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
-                assert_equal(len(self.nodes[0].getrawmempool()), size)
-                self.log.info('All of the recently mined transactions have been re-added into the mempool in {} seconds.'.format(end - start))
+                self.log.info('{}'.format(block_ids))
+                self.log.info('Block height: {}'.format(math.ceil(i / n_tx_to_mine)))
```
Yeah, not sure what I thinking here. Could've just use len(block_ids) as well. Removed in ac8b08b

reemuru commented:
Updated with ac8b08b. Let me know if I missed anything! Also, should I roll theses commits up into one, or will they get squashed at merge time? Forgot to sign the first few. 
Recent changes:
* `self.wallet.generate(1)` to `self.wallet.scan_blocks(start=76, num=1)` 
* remove blocks mined `block_ids` from cluttering logs along with unnecessary math import
* put back `enumerate` for mempool properties checks
* added assert before mining block

DariusParvin commented:
The changes look good to me! I'm relatively new myself but I think at this stage it would make sense to [squash](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits) everything into one commit.

reemuru commented:
> The changes look good to me! I'm relatively new myself but I think at this stage it would make sense to [squash](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits) everything into one commit.

Got it, thank you for your assistance!

sriramdvt commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +34,49 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
         More details: https://en.wikipedia.org/wiki/Tournament_(graph_theory)
         """
```
The old implementation of the test explicitly sets the ancestors and descendants of each transaction to form an acyclic tournament. However, `wallet.send_self_transfer()` only explicitly sets one ancestor for each transaction to form something like a singly linked list of `size` number of transactions. The acyclic tournament is then formed by the mempool. It would perhaps be a good idea to mention this in the description of `transaction_graph_test`

reemuru replied:
Updated the comment.

sriramdvt commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +34,49 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
+        block_ids = []
+        tx_size = []
+        for i in range(size):
+            txs.append(self.wallet.send_self_transfer(from_node=node))
+            tx_size.append(96)
```
In the earlier implementation of MiniWallet, `tx_size` had only one fixed value. To account for the varying sizes of transactions in the MiniWallet, it would be a better idea to calculate the transaction size from the transaction returned by `send_self_transfer`.

sriramdvt commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -38,86 +34,49 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
+        node.invalidateblock(block_ids[0])
+        end = time.time()
+        self.log.info('Mempool size post-invalidation: {}'.format(len(node.getrawmempool())))
+        assert_equal(len(self.nodes[0].getrawmempool()), size)
```
nit: Replace `self.nodes[0]` with `node`

sriramdvt reviewed (changes requested):
tACK df07919caa36d1345590010f5552d9e24db2bbde. The test works properly and checks what it is supposed to check. I strongly recommend making these changes for clarity and to maintain forward compatibility with MiniWallet.

reemuru reviewed (commented):
> tACK [df07919](https://github.com/bitcoin/bitcoin/commit/df07919caa36d1345590010f5552d9e24db2bbde). The test works properly and checks what it is supposed to check. I strongly recommend making these changes for clarity and to maintain forward compatibility with MiniWallet.

Thanks for reviewing! Made some changes based on your suggestions.

josibake commented:
Concept ACK

Thanks for posting the profiling data! Definitely agree with more miniwallet usage, especially if it increases test execution speed. I compiled and ran the tests and everything passed, but I would definitely recommend running `flake8` and making the suggested style changes. you can do this automatically with `autopep8` (I did it by with the following steps):

```bash
pip install autopep8
autopep8 --in-place mempool_updatefromblock.py
```
you can double check the changes by running `flake8 mempool_updatefromblock.py`. you should only see line length warnings, which are fine to ignore

practicalswift commented:
Concept ACK

reemuru commented:
> ```shell
> pip install autopep8
> ```

oh shoots! never used autopep8 before. will use from now on. Thanks for the tip! (^_^)

josibake commented:
> > ```shell
> > pip install autopep8
> > ```
> 
> oh shoots! never used autopep8 before. will use from now on. Thanks for the tip! (^_^) => [0b7dbeb](https://github.com/bitcoin/bitcoin/commit/0b7dbebd14962080c86c5667a77bbafdcefa6865)

looks great! id suggest squashing into one atomic commit per [the contributing guide](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits)

reemuru commented:
> > > ```shell
> > > pip install autopep8
> > > ```
> > 
> > 
> > oh shoots! never used autopep8 before. will use from now on. Thanks for the tip! (^_^) => [0b7dbeb](https://github.com/bitcoin/bitcoin/commit/0b7dbebd14962080c86c5667a77bbafdcefa6865)
> 
> looks great! id suggest squashing into one atomic commit per [the contributing guide](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits)

Got it, thanks much!

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/21999/commits/446c06a6fd8bf205e534be47c02d8088e40dd10f

built without wallet, ran the test to verify everything is behaving as expected. also code reviewed to verify the logic of the test is unchanged (strictly a refactor)

MarcoFalke commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -36,87 +34,65 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
-            assert_equal(self.nodes[0].getrawmempool(True)[tx]['ancestorcount'], k + 1)
-            assert_equal(self.nodes[0].getrawmempool(True)[tx]['ancestorsize'], sum(tx_size[0:(k + 1)]))
+            assert_equal(node.getrawmempool(True)[
+                         id]['descendantcount'], size - k)
```
please don't commit large style changes in the same commit as refactors/features. This makes review harder because it is not clear what is a refactor/style-change/feature

MarcoFalke replied:
Also I don't think the change here makes the code easier to read (neither for devs nor for scripts and parsers)

reemuru replied:
Oh shoot, my bad. :sob:  Was using some auto-styling tool and must've finger fudged it. Fixed and squashed into 33865a5

josibake replied:
> Oh shoot, my bad. sob Was using some auto-styling tool and must've finger fudged it. Fixed and squashed into [33865a5](https://github.com/bitcoin/bitcoin/commit/33865a5fbacf462c90520cb8f5dde61dd8fdbfc9)

my fault, @reemuru , i gave bad advice. @MarcoFalke is correct: linting (style changes) should be in separate PR's from refactors/features

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/22330/commits/c1c0768b619ab463a00052d0a584fba191eb02a9

built without wallet and ran test

MarcoFalke commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -9,20 +9,18 @@
         self.num_nodes = 1
-        self.extra_args = [['-limitdescendantsize=1000', '-limitancestorsize=1000']]
+        self.extra_args = [
+            ['-limitdescendantsize=1000', '-limitancestorsize=1000']]
```
here are still style changes mixed in

MarcoFalke commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -36,87 +34,62 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
-        self.transaction_graph_test(size=100, n_tx_to_mine=[25, 50, 75])
+        # Use batch size limited by DEFAULT_ANCESTOR_LIMIT = 25
+        # to not fire "too many unconfirmed parents" error.
+        self.transaction_graph_test(size=100, n_tx_to_mine=25)
```
why is this changed?

MarcoFalke replied:
The goal should probably be to fix the error, not to remove iterations from the test. I guess you can create more outputs or mine a block or something?

reemuru replied:
Ah ok, apologies for the noise. Will put this into draft mode since some work is still needed. Thanks for review!

reemuru reviewed (commented):
Ah ok, apologies for the noise. Will put this into draft mode since some work is still needed. Thanks for review!

adamjonas commented:
@reemuru was there a reason this closed? Would you like to mark it up for grabs?

pg156 commented:
I will start working on this as it is up for grabs. Please let me know if this is done elsewhere or no longer necessary?

MarcoFalke commented:
You can open the file on the `master` branch and check if `skip_if_no_wallet` is still in the file to see if this is still relevant.

pg156 commented:
Thanks @MarcoFalke. I see `skip_if_no_wallet` is still in https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_updatefromblock.py. I also checked each revision to the file since this PR was created, and don't see any change making this PR irrelevant. Did I miss anything before determining it is useful to work on this?
